### PR TITLE
[XGrid] Fix server-side multi filters

### LINKED ISSUE #1026: [XGrid] Server-side filtering and multi-column filtering don't work well together
- [x] The issue is present in the latest release.
- [x] I have searched the [issues](https://github.com/mui-org/material-ui/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

When running server-side filtering (eg the example given [here](https://material-ui.com/components/data-grid/filtering/#server-side-filter)) but changed to XGrid),  if you press the "add filter" button in the panel it seems like nothing happens. No additional filter row is added as expected (and how it works in XGrid without server-side filtering). 
Screen Shot 2021-02-10 at 12 38 15 PM

Clicking elsewhere and then back to filter loads the filter panel again, it will now have N rows for every time you clicked "add filter" before
Screen Shot 2021-02-10 at 12 38 23 PM

## Expected Behavior 🤔

When pressing "add filter" while running server-side filtering in XGrid, an additional row should appear in the filter panel

## Steps to Reproduce 🕹

Steps:

1. Load https://material-ui.com/components/data-grid/filtering/#server-side-filter locally or in codesandbox, using XGrid instead of DataGrid – example here: https://codesandbox.io/s/material-demo-forked-mv2xz?file=/demo.js
2. Filter the column
3. Press the "add filter" button
4. Observe no change to filter panel until you click elsewhere then back on filters

## Context 🔦

A user should be able to both use server-side filtering and multi-column filtering nicely together

## Your Environment 🌎
Observed using Chrome 88
  
```
 System:
    OS: macOS 11.1
  Binaries:
    Node: 15.8.0 - /usr/local/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 7.5.0 - /usr/local/bin/npm
  Browsers:
    Chrome: 88.0.4324.150
    Edge: Not Found
    Firefox: 84.0.1
    Safari: 14.0.2
  npmPackages:
    @material-ui/core: ^4.11.3 => 4.11.3 
    @material-ui/icons: ^4.11.2 => 4.11.2 
    @material-ui/lab: ^4.0.0-alpha.57 => 4.0.0-alpha.57 
    @material-ui/styles:  4.11.3 
    @material-ui/system:  4.11.3 
    @material-ui/types:  5.1.0 
    @material-ui/utils:  4.11.2 
    @material-ui/x-grid: ^4.0.0-alpha.19 => 4.0.0-alpha.19 
    @material-ui/x-grid-data-generator: ^4.0.0-alpha.19 => 4.0.0-alpha.19 
    @material-ui/x-license:  4.0.0-alpha.19 
    @types/react:  17.0.1 
    react: ^17.0.1 => 17.0.1 
    react-dom: ^17.0.1 => 17.0.1 

```
</details>

### PR DESCRIPTION
**Pull request by dtassone**
fix #1026

### CODE REVIEW

File: packages/storybook/src/stories/grid-filter.stories.tsx [hunk 1]
```diff
@@ -571,10 +567,12 @@ const filterModel = {
...
+  const [columns, setColumns] = React.useState(data.columns);
 
   React.useEffect(() => {
     if (data.columns.length > 0) {
-      const ratingColumn = data.columns.find((col) => col.field === 'rating');
+      let newColumns = [...data.columns];
+      const ratingColumn = {...newColumns.find((col) => col.field === 'rating')};
 
       const ratingOperators = getNumericColumnOperators();
       ratingColumn!.filterOperators = ratingOperators.map((operator) => {
```
Comment (dtassone) [inline packages/storybook/src/stories/grid-filter.stories.tsx:578]: All filters stories that changes the columns are broken due #979

**Comment by dtassone** [comment]
waiting for #1024

File: packages/grid/_modules_/grid/components/panel/filterPanel/FilterPanel.tsx [hunk 2]
```diff
@@ -70,7 +70,12 @@ export function FilterPanel() {
       </PanelContent>
       {!disableMultipleColumnsFiltering && (
         <PanelFooter>
-          <Button onClick={addNewFilter} startIcon={<AddIcon />} color="primary">
+          <Button
+            onClick={addNewFilter}
+            startIcon={<AddIcon />}
+            color="primary"
+            className="MuiAddFilterButton"
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/panel/filterPanel/FilterPanel.tsx:77]:
```suggestion
```

File: packages/grid/x-grid/src/tests/filtering.XGrid.test.tsx [hunk 3]
```diff
@@ -224,4 +226,85 @@ describe('<XGrid /> - Filter', () => {
...
+              state={{
+                preferencePanel: { open: true, openedPanelValue: PreferencePanelsValue.filters },
+              }}
+            />
+          </div>
+        );
+      }
+
+      render(<AddServerFilterGrid />);
+      const addButton = document.querySelector(`.MuiAddFilterButton`);
```
Comment (oliviertassinari) [inline packages/grid/x-grid/src/tests/filtering.XGrid.test.tsx:303]: use the accessibility tree

File: packages/grid/x-grid/src/tests/filtering.XGrid.test.tsx [hunk 4]
```diff
@@ -1,20 +1,22 @@
-import { expect } from 'chai';
-import * as React from 'react';
-import { useFakeTimers } from 'sinon';
-import { getColumnHeaderCell, getColumnValues } from 'test/utils/helperFn';
-import {
-  createClientRenderStrictMode,
-  // @ts-expect-error need to migrate helpers to TypeScript
-  fireEvent,
-} from 'test/utils';
```
Comment (oliviertassinari) [inline packages/grid/x-grid/src/tests/filtering.XGrid.test.tsx:9]: Why this diff? Could we minimize them?

On a related note, I believe that so far, the approach is to import the most common modules first, e.g. react, then sinon, then test utils, then the module we test.

File: packages/grid/x-grid/src/tests/filtering.XGrid.test.tsx [hunk 5]
```diff
@@ -224,4 +226,85 @@ describe('<XGrid /> - Filter', () => {
...
+          </div>
+        );
+      }
+
+      render(<AddServerFilterGrid />);
+      const addButton = document.querySelector(`.MuiAddFilterButton`);
+      clock.tick(100);
+      fireEvent.click(addButton);
+      const filterForms = document.querySelectorAll(`.MuiDataGridFilterForm-root`);
+      expect(filterForms.length).to.eq(2);
```
Comment (oliviertassinari) [inline packages/grid/x-grid/src/tests/filtering.XGrid.test.tsx:307]: We should likely make this alias throw in the setup

```suggestion
      expect(filterForms.length).to.equal(2);
```

or maybe better

```suggestion
      expect(filterForms).to.have.length(2);
```

File: packages/grid/x-grid/src/tests/filtering.XGrid.test.tsx [hunk 6]
```diff
@@ -224,4 +226,85 @@ describe('<XGrid /> - Filter', () => {
...
+            if (!commodityFilterValue) {
+              resolve(serverRows);
+              return;
+            }
+            resolve(
+              serverRows.filter(
+                (row) => row.commodity.toLowerCase().indexOf(commodityFilterValue) > -1,
+              ),
+            );
+          }, Math.random() * 500 + 100); // simulate network latency
```
Comment (oliviertassinari) [inline packages/grid/x-grid/src/tests/filtering.XGrid.test.tsx:252]: This Math.random() and setTimeout are red flags. Please simplify the test case as much as possible to nail the root cause (failing on master). We don't need the ceremony.

**Review by oliviertassinari — CHANGES_REQUESTED** [review]

**Review by oliviertassinari — APPROVED** [review]

**Comment by oliviertassinari** [comment]
I have pushed 3 commits to simplify the tests

**Review by DanailH — APPROVED** [review]

**Pull request merged**